lib/Basics/process-utils.cpp

```diff
@@ -225,9 +225,115 @@ static bool CreatePipes(int* pipe_server_to_child, int* pipe_child_to_server) {
 /// @brief starts external process
 ////////////////////////////////////////////////////////////////////////////////
 
-static void StartExternalProcess(ExternalProcess* external, bool usePipes,
-                                 std::vector<std::string> const& additionalEnv,
-                                 std::string const& fileForStdErr) {
+#include <spawn.h>
+
+static void StartExternalProcessPosixSpawn(
+    ExternalProcess* external, bool usePipes,
+    std::vector<std::string> const& additionalEnv,
+    std::string const& fileForStdErr) {
+  int pipe_server_to_child[2];
+  int pipe_child_to_server[2];
+
+  if (usePipes) {
+    bool ok = CreatePipes(pipe_server_to_child, pipe_child_to_server);
+
+    if (!ok) {
+      external->_status = TRI_EXT_PIPE_FAILED;
+      return;
+    }
+  }
+
+  posix_spawn_file_actions_t file_actions;
+  posix_spawn_file_actions_init(&file_actions);
+  // file actions are performed in order they were added.
+
+  if (usePipes) {
+    posix_spawn_file_actions_adddup2(&file_actions, pipe_server_to_child[0], 0);
+    posix_spawn_file_actions_adddup2(&file_actions, pipe_child_to_server[1], 1);
+
+    posix_spawn_file_actions_addclose(&file_actions, pipe_server_to_child[0]);
+    posix_spawn_file_actions_addclose(&file_actions, pipe_server_to_child[1]);
+    posix_spawn_file_actions_addclose(&file_actions, pipe_child_to_server[0]);
+    posix_spawn_file_actions_addclose(&file_actions, pipe_child_to_server[1]);
+  } else {
+    posix_spawn_file_actions_addopen(&file_actions, 0, "/dev/null", O_RDONLY,
+                                     0);
+  }
+
+  if (!fileForStdErr.empty()) {
+    posix_spawn_file_actions_addopen(&file_actions, 2, fileForStdErr.c_str(),
+                                     O_CREAT | O_WRONLY | O_TRUNC, 0644);
+  }
+
+  posix_spawnattr_t spawn_attrs;
+  posix_spawnattr_init(&spawn_attrs);
+
+  std::vector<char*> envs;
+  for (char** e = environ; *e != nullptr; ++e) {
+    envs.push_back(*e);
+  }
+
+  envs.reserve(additionalEnv.size() + 1);
+  std::transform(
+      additionalEnv.begin(), additionalEnv.end(), std::back_inserter(envs),
+      [](auto& str) -> char* { return const_cast<char*>(str.data()); });
+  envs.emplace_back(nullptr);
+
+  int result = posix_spawnp(&external->_pid, external->_executable.c_str(),
```

Comment:
I am not 100% clear about how signal handling for the child process will work with this call.
The previous code that uses `fork` and `exec` called `arangodb::signals::unmaskAllSignals();` for the child, to unblock all signals.
If this is still happening by default when using `posix_spawnp` is unknown to me. I read the man page (https://man7.org/linux/man-pages/man3/posix_spawn.3p.html) but found it confusing:

        If the POSIX_SPAWN_SETSIGMASK flag is set in the spawn-flags
       attribute of the object referenced by attrp, the child process
       shall initially have the signal mask specified in the spawn-
       sigmask attribute of the object referenced by attrp.

       If the POSIX_SPAWN_SETSIGDEF flag is set in the spawn-flags
       attribute of the object referenced by attrp, the signals
       specified in the spawn-sigdefault attribute of the same object
       shall be set to their default actions in the child process.
       Signals set to the default action in the parent process shall be
       set to the default action in the child process.

       Signals set to be caught by the calling process shall be set to
       the default action in the child process.

       Except for SIGCHLD, signals set to be ignored by the calling
       process image shall be set to be ignored by the child process,
       unless otherwise specified by the POSIX_SPAWN_SETSIGDEF flag
       being set in the spawn-flags attribute of the object referenced
       by attrp and the signals being indicated in the spawn-sigdefault
       attribute of the object referenced by attrp.

       If the SIGCHLD signal is set to be ignored by the calling
       process, it is unspecified whether the SIGCHLD signal is set to
       be ignored or to the default action in the child process, unless
       otherwise specified by the POSIX_SPAWN_SETSIGDEF flag being set
       in the spawn_flags attribute of the object referenced by attrp
       and the SIGCHLD signal being indicated in the spawn_sigdefault
       attribute of the object referenced by attrp.

Reply:
I have added code to handle this.,